File: Data Structures & Algorithms/valid-parenthesis-string/submission-3.java
class Solution {
    public boolean checkValidString(String s) {
        int left = 0;
        int right = 0;
        for (char c : s.toCharArray()) {
            if (c == '('){
                left ++;
                right ++;
            }
            else if (c == ')'){
                left --;
                right--;

            }
            else {
                left --;
                right ++;
            }
            if (right < 0)
                return false;
            left = Math.max(0, left);
        }
        return left == 0;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-3.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int sol{};
        int left{1};
        int maxi{0};
        for (const auto& p : piles) {
            maxi = std::max(maxi, p);
        }
        int right{maxi};
        while (left <= right) {
            int mid{left + (right - left) / 2};
            int hours{};
            for (const auto& p : piles) {
                hours += p / mid;
                hours += ((p % mid != 0) ? 1 : 0);
            }
            if (hours > h) {
                left = mid + 1;
            }
            else {
                sol = mid;
                right = mid - 1;
            }
        }
        return sol;
    }
};
